src/governance.cpp

```diff
@@ -550,15 +525,39 @@ void CGovernanceManager::UpdateCachesAndClean()
                     ++lit;
                 }
             }
-            if(pObj->nObjectType == GOVERNANCE_OBJECT_WATCHDOG) {
-                mapWatchdogObjects.erase(it->first);
+
+            int nStatus;
+
+            int64_t nSuperblockCycleSeconds = Params().GetConsensus().nSuperblockCycle * Params().GetConsensus().nPowTargetSpacing;
+            int64_t nTimeExpired = pObj->GetCreationTime() + 2 * nSuperblockCycleSeconds + GOVERNANCE_DELETION_DELAY;
+
+            if(pObj->GetObjectType() == GOVERNANCE_OBJECT_WATCHDOG) {
+                mapWatchdogObjects.erase(nHash);
+                nStatus = (int)nTimeExpired;
+            } else if(pObj->GetObjectType() == GOVERNANCE_OBJECT_TRIGGER) {
+                nStatus = (int)nTimeExpired;
+            } else {
+                // keep hashes of deleted proposals forever
+                nStatus = std::numeric_limits<int>::max();
             }
+
+            mapSeenGovernanceObjects.insert(std::make_pair(nHash, nStatus));
             mapObjects.erase(it++);
         } else {
             ++it;
         }
     }
 
+    // forget about expired deleted objects
```

Comment:
This is relevant for proposals only though, right? Because wds and triggers have expiration rules (unlike expired/deleted proposals which can be removed from the system only via votes atm). This can probably be fixed via validation `end_epoch` in data, which however goes somewhat against the principle that dashd should be data-agnostic. Another way to fix this would be to include a "period of validity" i.e. start/end block (or timestamp) to gov obj structure and validate it appropriately.